File: app/api/v1/reports.py
from datetime import date

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.dependencies import get_current_user
from app.models.user import User
from app.services.report_service import ReportService

router = APIRouter()


@router.get("/spending-by-category")
async def spending_by_category(
    date_from: date = Query(...),
    date_to: date = Query(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await ReportService(db).spending_by_category(current_user.tenant_id, date_from, date_to)


@router.get("/income-vs-expenses")
async def income_vs_expenses(
    date_from: date = Query(...),
    date_to: date = Query(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await ReportService(db).income_vs_expenses(current_user.tenant_id, date_from, date_to)


@router.get("/net-worth")
async def net_worth(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await ReportService(db).net_worth(current_user.tenant_id)


@router.get("/budget-vs-actual")
async def budget_vs_actual(
    date_from: date = Query(...),
    date_to: date = Query(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await ReportService(db).budget_vs_actual(current_user.tenant_id, date_from, date_to)

File: frontend/src/api/reports.ts
import apiClient from './client';
import type { BudgetProgress, IncomeVsExpenses, NetWorth, SpendingByCategory } from '../types';

export const reportsApi = {
  spendingByCategory: async (date_from: string, date_to: string): Promise<SpendingByCategory[]> => {
    const { data } = await apiClient.get<SpendingByCategory[]>('/reports/spending-by-category', {
      params: { date_from, date_to },
    });
    return data;
  },

  incomeVsExpenses: async (date_from: string, date_to: string): Promise<IncomeVsExpenses> => {
    const { data } = await apiClient.get<IncomeVsExpenses>('/reports/income-vs-expenses', {
      params: { date_from, date_to },
    });
    return data;
  },

  netWorth: async (): Promise<NetWorth> => {
    const { data } = await apiClient.get<NetWorth>('/reports/net-worth');
    return data;
  },

  budgetVsActual: async (date_from: string, date_to: string): Promise<BudgetProgress[]> => {
    const { data } = await apiClient.get<BudgetProgress[]>('/reports/budget-vs-actual', {
      params: { date_from, date_to },
    });
    return data;
  },
};
